tempesta_fw/http_parser.c

```diff
@@ -457,6 +457,134 @@ parse_int_hex(unsigned char *data, size_t len, unsigned long *acc)
 	return CSTR_POSTPONE;
 }
 
+/**
+ * Add unconditionaly hop-by-hop headers to list of hop-by-hop headers names.
+ * Do not init raw headers table until we found that Connection header contains
+ * atleast one raw header.
+ */
+static void
+__hbh_parser_init(TfwHttpHbhHdrs *hbh_hdrs, bool req)
+{
+	/*
+	 * Connection header is hop-by-hop header.
+	 *
+	 * Server header isn't defined as hop-by-hop by the RFC, but we
+	 * don't show protected server to world. User-agent and Server headers
+	 * have the same id in tfw_http_hdr_t, so Server header must be marked
+	 * only for responses.
+	 */
+	hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_CONNECTION;
+	if (!req)
+		hbh_hdrs->spec |= 0x1 << TFW_HTTP_HDR_SERVER;
+}
+
+#define MARK_SPEC_HBH(hid)						\
+	if ((msg->parser.hbh_parser.spec & (0x1 << hid))		\
+	    && !TFW_STR_EMPTY(&msg->parser.hdr))			\
+		msg->parser.hdr.flags |= TFW_STR_HBH_HDR;
+
+/**
+ * Mark raaw header @hdr as hop-by-hop if its name was listed in Connection
+ * header
+ */
+static void
+mark_raw_hbh(TfwHttpMsg *hm, TfwStr *hdr)
+{
+	TfwHttpHdrTbl *ht = hm->parser.hbh_parser.raw;
+	unsigned int i;
+
+	if (!ht)
+		return;
+
+	for (i = 0; i < ht->off; ++i) {
+		if (!tfw_stricmpspn(&ht->tbl[i], hdr, ':')) {
+			hdr->flags |= TFW_STR_HBH_HDR;
+			break;
+		}
+	}
+}
+
+static int
+__hbh_parser_alloc_table(TfwHttpMsg *hm)
+{
+	hm->parser.hbh_parser.raw = (TfwHttpHdrTbl *)
+			tfw_pool_alloc(hm->pool, TFW_HHTBL_SZ(1));
+	if (unlikely(!hm->parser.hbh_parser.raw)) {
+		return -ENOMEM;
+	}
+
+	hm->parser.hbh_parser.raw->size = __HHTBL_SZ(1);
+	hm->parser.hbh_parser.raw->off = 1;
+	memset(hm->parser.hbh_parser.raw->tbl, 0,
+	       __HHTBL_SZ(1) * sizeof(TfwStr));
+
+	return 0;
+}
+
+/**
+ * Add header name listed in Connection header to rable of raw headers.
+ * If @last is true then (@data, @len) represnts last chunk of header name and
+ * chunk with ':' will be added to the end. Otherwize last header in table stays
+ * open to add more data.
+ *
+ * After name of hop-by-hop header was completed, we will search for headers
+ * with that name and mark them as hop-by-hop.
+ */
+static int
+__hbb_parser_add_data(TfwHttpMsg *hm, char *data, unsigned long len, bool last)
+{
+	unsigned int hid;
+	TfwStr *hdr, *append;
+	int r;
+	TfwHttpHdrTbl *ht = hm->parser.hbh_parser.raw;
+	if (!ht) {
+		if ((r = __hbh_parser_alloc_table(hm)))
+			return r;
+		ht = hm->parser.hbh_parser.raw;
+	}
+
+	if (ht->tbl[ht->off - 1].flags & TFW_STR_COMPLETE) {
+		hid = ht->off;
+		if (hid == ht->size)
+			if (__tfw_http_msg_grow_hdr_tbl(&hm->parser.hbh_parser.raw,
+							hm->pool))
+				return -ENOMEM;
```

Comment:
This is true and powerful headers parser. Do we really allow that Connection mentioning more than 16 header? I believe it has sense to block the header. Meantime it will make the code faster and easier and that's good.